### ISSUE TITLE
to_tensor and pil_to_tensor inconsistency

### ISSUE THREAD

**Issue by SsnL** [pytorch/vision#2269]
`to_tensor` uses `from_numpy` which never copies and always returns a CPU tensor. But `pil_to_tensor` uses `as_tensor`, which would return a CUDA tensor if the default tensor type is changed by the user (e.g., https://github.com/pytorch/pytorch/issues/39088). I think a discussion on what the correct behavior is is needed, followed with a patch to make the two functions consistent.

**Comment by fmassa** [comment]
Hi Simon,

That's a good question, and I had a quick look at the linked discussion to get some more context.

In principle, I think that the current design of the DataLoader with multiprocessing is not ideal to support tensors being passed on the GPU (but I might be wrong).

This makes me think that we might want to enforce in torchvision that `pil_to_tensor` always returns a CPU tensor, but that can also be contradicting the contract of setting the global default device to be CUDA.

Thoughts?

**Comment by SsnL** [comment]
I agree what the DataLoader is best suited to load CPU tensors. However, loading CUDA tensor is one of the things we support too, and unfortunately it seems that the DataLoader is one of the easier ways in PyTorch to do so. 

However, I am hesitant to say that these methods should respect `default_tensor_type`, for potential BC reasons as `to_tensor` has existed for a long time. One common case should be that user (on Linux) sets `default_tensor_type` to CUDA and the forked DataLoader workers calls `ToTensor()`. 

Maybe we can specify in docs that these methods / classes move to CPU by default, and introduce an `device=` argument? :)

**Comment by ain-soph** [comment]
Personal opinion is to have a quick fix by forcing ``cpu`` tensors, which is a bug fix. And things will work without exceptions. Of course it might be specified in the documents.

The capability to load ``gpu`` tensors for multiprocessing could be discussed as new features, since it seems to require more efforts.

**Comment by SsnL** [comment]
Loading to GPU is and has always been supported. It’s not a new feature.

**Comment by ain-soph** [comment]
> Loading to GPU is and has always been supported. It’s not a new feature.

I'm not quite sure about that? Is their any way to use ``DataLoader`` to load data to GPU? Currently, when I set ``num_workers=0`` or ``spawn`` start method, I still get CPU tensors based on what you have illustrated. 

If I use ``Dataloader``, I need to transform them to ``cuda`` manually. If I want to load data directly to gpu, I need to code the pipeline in ``dali`` by myself.
I might be not familiar with new pytorch features. Is there any update about that? It'll be a great news to know it's supported.
I see that the imagenet example still need to use ``cuda()`` to transform data to gpu, which is originally loaded from ``DataLoader``.
https://github.com/pytorch/examples/blob/master/imagenet/main.py#L283-L284

**Comment by SsnL** [comment]
It’s been in pytorch since a long time ago. Probably before 1.0 if I remember correctly. As long as you return a CUDA tensor from your dataset code and use proper start method, the main process will receive a CUDA tensor, regardless of num_workers. 

It is always not advertised nor recommended because it makes CUDA memory management difficulty, moving to GPU really fast and not usually the bottleneck as long as you use pin_memroy. 

This issue is not about whether pytorch supports loading CUDA tensors. It is about torchvision, a library built on top of pytorch, should return CUDA tensor or not in ToTensor that is called in the dataset code. Any dataset code that yields CUDA tensors has always been working fine.
